app/storage/encrypted_questionnaire_storage.py

```diff
@@ -1,49 +1,89 @@
+class EncryptedQuestionnaireStorage:
 
-    def __init__(self, user_id, user_ik):
-        super().__init__(user_id)
+    def __init__(self, user_id, user_ik, pepper):
+        if user_id is None:
+            raise ValueError('User id must be set')
         if user_ik is None:
             raise ValueError('User ik must be set')
+        if pepper is None:
+            raise ValueError('Pepper must be set')
         self.encryption = JWEDirEncrypter()
         self.decryption = JWEDirDecrypter()
-        self.user_ik = user_ik
+        self.user_id = user_id
+
+        sha256 = hashlib.sha256()
+        sha256.update(to_str(user_id).encode('utf-8'))
+        sha256.update(to_str(user_ik).encode('utf-8'))
+        sha256.update(to_str(pepper).encode('utf-8'))
+
+        # we only need the first 32 characters for the CEK
+        self.cek = sha256.hexdigest()[:32]
 
     def add_or_update(self, data):
         encrypted_data = self.encrypt_data(data)
-        super().add_or_update(encrypted_data)
+        questionnaire_state = self._get()
+        if questionnaire_state:
+            logger.debug("updating questionnaire data", user_id=self.user_id)
+            questionnaire_state.set_data(encrypted_data)
+        else:
+            logger.debug("creating questionnaire data", user_id=self.user_id)
+            questionnaire_state = QuestionnaireState(self.user_id, encrypted_data)
+
+        with commit_or_rollback(db_session):
+            # pylint: disable=maybe-no-member
+            # session has a add function but it is wrapped in a session_scope which confuses pylint
+            db_session.add(questionnaire_state)
 
     def get_user_data(self):
-        data = super().get_user_data()
+        data = self._get_questionnaire_state()
         if data is not None and 'data' in data:
-            decrypted_data = self.decrypt_data(self.user_id, self.user_ik, data)
+            decrypted_data = self.decrypt_data(data)
             return decrypted_data
 
+    def delete(self):
+        logger.debug("deleting users data", user_id=self.user_id)
+        questionnaire_state = self._get()
+        if questionnaire_state:
+            with commit_or_rollback(db_session):
+                # pylint: disable=maybe-no-member
+                # session has a delete function but it is wrapped in a session_scope which confuses pylint
+                db_session.delete(questionnaire_state)
+
     def encrypt_data(self, data):
-        sha_key = generate_key(self.user_id, self.user_ik)
+        sha_key = self.generate_key()
         encrypted = self.encryption.encrypt(data, sha_key)
         return {'data': encrypted}
 
-    def decrypt_data(self, user_id, user_ik, encrypted_data):
-        sha_key = generate_key(user_id, user_ik)
+    def decrypt_data(self, encrypted_data):
+        sha_key = self.generate_key()
         return self.decryption.decrypt(encrypted_data['data'], sha_key)
+
+    def generate_key(self):
```

Comment:
I'm reluctant to get rid of the `generate_key` method altogether since we have test coverage of this method that will tell us if the key generation is broken, which I think would be good to keep. I'll move the logic back into generate_key and call from __init__.